src/Adapter/CartRule/Repository/CartRuleRepository.php

```diff
@@ -48,10 +50,27 @@ public function __construct(
     ) {
     }
 
-    public function add(CartRule $cartRule): CartRule
+    /**
+     * @param CartRule $cartRule
+     * @param ShopId[] $associatedShopIds
+     *
+     * @return CartRule
+     */
+    public function add(CartRule $cartRule, array $associatedShopIds): CartRule
     {
+        // cart rule works well with cart_rule_shop table as all the other entities when we add the table association
+        // except it uses "shop_restrictions" property for some reason, so we force it to true to avoid breaking legacy code
+        //@todo: need to confirm. Is it ok that we always force the shop_restriction = true?
+        //       in legacy code shop_restriction = false means rule can be used in all shops
+        //       (and it is same if we add all the shops into cart_rule_shop table)
+        //       so can we just make sure new code always puts related shops into cart_rule_shop table (even if its single shop context)
+        //       and eventually we can get rid of shop_restriction prop?
+        //       also the shopTableAssociaton could be moved to Shop:init(), so we don't need to do it in handlers anymore (not sure how would it affect legacy code)
```

Comment:
ping @jolelievre.
shop_restriction is not used that much, I could get rid of all usages and deprecate it.. :arrow_up: :thinking:

Reply:
for now it should work this way as it did before, but it would be cleaner if we get rid of it completeley, and bring back that shop_association to cart_rule_shop by default (so it acts as all the other shop associations and not some "specific" shop restriction)

Reply:
https://github.com/PrestaShop/PrestaShop/pull/33050/commits/4f8e8bb9a33a3f407eebbb0ef3ed69404657c246
this seems to do the trick

Reply:
I agree with you that CartRule shop associations should be built the same way as for other entities. But just to be sure about the current behaviour, is there another table dedicated to shop restrictions? Or does it use `cart_rule_shop` and simply "pretends" like it's a restriction specific for cart rule when it's actually the same as the generic behaviour?

Regarding the `shop_restriction` and the fact that we should only rely on associated shops I agree with the logic, but then we must make sure that it's not used at all anymore anywhere in the code especially in the FO, then we can deprecate it and it will be removed in 10.0

Even all the other restriction fields kind of fell overkill, I guess it's a small cache to prevent doing a join request but I'm not sure it's really worth it. Anyway it's another topic!

Reply:
If we go with this new strategy then the shop IDs must be required in the creation command, but you already handled that!